src/x/hstreamdb-erl-nifs/src/lib.rs

```diff
@@ -190,72 +187,80 @@ fn stop_producer(producer: ResourceArc<NifAppender>) -> Atom {
-) -> Result<ResourceArc<AppendResultFuture>, Term<'a>> {
+) -> Result<(), Term<'a>> {
     let raw_payload = rustler::Binary::from_term(raw_payload)
-        .map_err(|err| (badarg(), format!("{err:?}")).encode(env))?;
-    let record = Record {
-        partition_key,
-        payload: hstreamdb::Payload::RawRecord(raw_payload.to_vec()),
+        .map_err(|err| (badarg(), format!("{err:?}")).encode(env))?
+        .to_vec();
+    let future = async move {
+        let record = Record {
+            partition_key,
+            payload: hstreamdb::Payload::RawRecord(raw_payload),
+        };
+        let producer = &producer.0;
+        if producer.send_async(Some((record, pid))).await.is_err() {
+            let mut env = OwnedEnv::new();
+            env.send_and_clear(&pid, |env| {
+                (append_reply(), error(), terminated()).encode(env)
+            })
+        }
     };
-    let producer = &producer.0;
-    let (sender, receiver) = oneshot::channel();
-    producer
-        .send(Some((record, sender)))
-        .map_err(|_| terminated().encode(env))?;
-    let receiver = receiver.blocking_recv().unwrap();
-    Ok(ResourceArc::new(AppendResultFuture(
-        Mutex::new(Some(receiver)),
-        OnceCell::new(),
-    )))
+    runtime::spawn(future);
+    Ok(())
 }
 
 #[rustler::nif]
-fn append<'a>(
+fn async_append<'a>(
     env: Env<'a>,
+    pid: LocalPid,
     producer: ResourceArc<NifAppender>,
     partition_key: String,
     raw_payload: Term,
 ) -> Term<'a> {
-    match try_append(env, producer, partition_key, raw_payload) {
-        Ok(x) => (ok(), x).encode(env),
+    match try_append(env, pid, producer, partition_key, raw_payload) {
+        Ok(()) => ok().encode(env),
         Err(err) => (error(), err).encode(env),
     }
 }
 
 #[rustler::nif]
-fn await_append_result(env: Env, x: ResourceArc<AppendResultFuture>) -> Term {
+fn async_await_append_result(pid: LocalPid, x: ResourceArc<AppendResultFuture>) {
     use crate::AppendResult::*;
-    let result = &x.1;
-
-    if result.get().is_none() {
-        let receiver: &Mutex<_> = &x.0;
-        let mut receiver: MutexGuard<Option<_>> = receiver.lock();
-        let receiver = mem::take(&mut (*receiver));
-        let append_result: Result<hstreamdb::RecordId, Arc<_>> =
-            receiver.unwrap().blocking_recv().unwrap();
-        let append_result = match append_result {
-            Ok(record_id) => RecordId(record_id),
-            Err(err) => Error(err.to_string()),
-        };
-        result.set(append_result).unwrap()
-    }
+    let future = async move {
+        let mut env = OwnedEnv::new();
+        let result = &x.1;
+        if result.get().is_none() {
+            let receiver: &Mutex<_> = &x.0;
+            let mut receiver: MutexGuard<Option<_>> = receiver.lock();
+            let receiver = mem::take(&mut (*receiver));
+            let append_result: Result<hstreamdb::RecordId, Arc<_>> =
+                receiver.unwrap().blocking_recv().unwrap();
```

Comment:
blocking in non blocking context

Reply:
[d317459](https://github.com/hstreamdb/hstreamdb-rust/pull/73/commits/d317459c12daba2a48d64a12297adafdae0fe58f)